### PR TITLE
FINERACT-838 Added Swagger Codegen Support

### PR DESCRIPTION
**Pull request by Grandolf49**
FINERACT-838

## Description
Integrated Swagger Codegen support for automatically generating client code from OpenAPI Specification

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Commit message starts with the issue number from https://issues.apache.org/jira/projects/FINERACT/. Ex: FINERACT-646 Pockets API.

- [x] Coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions have been followed.

- [x] API documentation at fineract-provider/src/main/resources/static/api-docs/apiLive.htm has been updated with details of any API changes.

- [x] Integration tests have been created/updated for verifying the changes made.

- [x] All Integrations tests are passing with the new commits.

- [x] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the list.)

Our guidelines for code reviews is at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide

### CODE REVIEW

**Comment by Grandolf49** [comment]
To generate the code:

1. **./gradlew build**
2. **./gradlew generateSwaggerCode**

The code will be generated in the directory **build/swagger-code-fineract**

To build the generated code:

1. Navigate to **build/swagger-code-fineract** directory
2. **./gradlew build**

A Retrofit2 based Java API Client for Fineract will be generated in build/swagger-code-fineract/build

To customize the generated code:

1. Navigate to [fineract-provider/config/swagger/config.json](https://github.com/Grandolf49/fineract/blob/swagger-codegen/fineract-provider/config/swagger/config.json)
2. Refer to [this](https://gist.github.com/Grandolf49/2d222c3d1d0b834ad8f02b5fde8b4c14) for available options in config.json

File: fineract-provider/config/swagger/config.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
+   "library": "retrofit2",
+   "modelPackage": "org.mifos.models",
```
Comment (vorburger) [inline fineract-provider/config/swagger/config.json]: We shouldn't use "org.mifos" for Java package names that are code generated in an Apache project... how about making this e.g. org.apache.fineract.client?

**Review by vorburger — CHANGES_REQUESTED** [review]
could you put the description about how to use this which you provided above into the/README.md as part of this PR?

File: fineract-provider/config/swagger/config.json (hunk 1 above)
Comment (Grandolf49) [reply fineract-provider/config/swagger/config.json]: Sorry about that! I missed it out. I'll change it to **org.apache.fineract.client**

To verify the changes:
- "modelPackage": "org.apache.fineract.client.models"
- "apiPackage": "org.apache.fineract.client.services"
- "invokerPackage": "org.apache.fineract.client"
- "groupId": "org.apache.fineract"
- "artifactId": "client"

This is how we want right?

> could you put the description about how to use this which you provided above into the/README.md as part of this PR?
Yeah sure. I'll add this in the Readme.md

**Comment by vorburger** [comment]
SGTM.

>

**Comment by vorburger** [comment]
> could you put the description about how to use this which you provided above into the/README.md as part of this PR?

@Grandolf49 actually, how about, similarly to the discussions in FINERACT-742 and FINERACT-834, you put the doc how to use this (which you provided in the comment above into) into... `docs/developers/swagger/client.md`? And then just a add 1 sentence linking to that from [README.md#apache-fineract-platform-api](https://github.com/apache/fineract/blob/develop/README.md#apache-fineract-platform-api) (both as part of this PR). That would be a great start, for more `docs/` to come, hopefully.

File: README.md [hunk 2]
```diff
@@ -263,6 +263,28 @@ API clients (Web UIs, Mobile, etc.)
 * https://github.com/openMF has more related proejcts
 
 
+Generate Apache Fineract API Client
```
Comment (Grandolf49) [inline README.md:266]: @vorburger Actually, I already added the description in the README.md. 
But I can definitely create a new file, as you suggested, under `docs/developers/swagger/client.md` and link it with the main README.md

File: README.md [hunk 3]
```diff
@@ -263,6 +263,28 @@ API clients (Web UIs, Mobile, etc.)
...
+============
+
+Apache Fineract supports client code generation using [Swagger Codegen](https://github.com/swagger-api/swagger-codegen). Project supports [all clients](https://github.com/swagger-api/swagger-codegen#overview) supported by Swagger Codegen. It uses [OpenAPI Specification Version 3.0.3](https://swagger.io/specification/).
+
+To generate code, execute the following commands:
+
+- **./gradlew build**
+- **./gradlew generateSwaggerCode**
+
+The client code will be generated in the directory **build/swagger-code-fineract**
```
Comment (vorburger) [inline README.md]:
```suggestion
The client code will be generated in the directory `build/swagger-code-fineract`.
```

File: README.md [hunk 4]
```diff
@@ -263,6 +263,28 @@ API clients (Web UIs, Mobile, etc.)
...
+
+To build the generated code:
+
+- Navigate to **build/swagger-code-fineract** directory
```
Comment (vorburger) [inline README.md]:
```suggestion
- Navigate to `build/swagger-code-fineract` directory (`cd`)
```

File: README.md [hunk 5]
```diff
@@ -263,6 +263,28 @@ API clients (Web UIs, Mobile, etc.)
...
+- **./gradlew build**
```
Comment (vorburger) [inline README.md]:
```suggestion
- Run `/gradlew build`
```

File: README.md [hunk 6]
```diff
@@ -263,6 +263,28 @@ API clients (Web UIs, Mobile, etc.)
...
+
+To customize the generated code:
+
+- Open **fineract-provider/config/swagger/config.json**
+- Refer to [this](https://gist.github.com/Grandolf49/2d222c3d1d0b834ad8f02b5fde8b4c14) for available options in config.json
```
Comment (vorburger) [inline README.md]: If you move it to a separate MD docs file, you could just "inline" your entire Gist? Better. Actually, perhaps even better you could link to the appropriate docs page of Swagger CodeGen where they document this? Assuming they have proper doc! :imp: That way, it stays up to date, as we'll upgrade the plugin in the future.

File: fineract-provider/config/swagger/config.json [hunk 7]
```diff
@@ -0,0 +1,11 @@
+{
+   "groupId": "org.apache.fineract",
+   "artifactId": "client",
+   "artifactVersion": "1.0.0",
```
Comment (vorburger) [inline fineract-provider/config/swagger/config.json]:
```suggestion
   "artifactVersion": "1.4.0",
```

File: README.md (hunk 6 above)
Comment (Grandolf49) [reply README.md]: I spent a lot of time trying to find where they have documented these options. 😅 [This](https://github.com/swagger-api/swagger-codegen/issues/7795) thread was useful but just a workaround.

I also came across [this](https://openapi-generator.tech/docs/generators/java) which seems like the official documentation from the OpenAPI Team. 

I think it's just better to move this to a separate doc. I'll put a link to this in the doc as you suggested

Comment (Grandolf49) [reply README.md]: @vorburger I have done the necessary changes and created a new client.md. I could also spot some minor corrections in the main readme.md file. So I have updated those too! I hope that's okay! 😅

Let me know if I need to change anything else.

**Review by vorburger — CHANGES_REQUESTED** [review]
@Grandolf49 the changes you are proposing to make to the README as part of this Swagger client contribution are (wrongly) changing the meaning of some of the paragraphs. Also your MD editor appears to auto format MD 1./1./1. as 1./2./3.? Both are holding up merging this PR. (Raise a few separate PR with only those changes, separate e.g. trivial formatting fixes VS 1./2./3. VS language change, if you would like feedback on why some those are problematic.)

Could I suggest that you revert most of those changes to the README, and only modify what you need for the Swagger client, the new part for the client.md? I'd love to be able to merge that ASAP! 

PS: We can further discuss on the linked JIRA what you could next, if you are interested in doing anything more here?

**Comment by Grandolf49** [comment]
Yeah, sure. I'll revert those changes and update my PR.

Would love to discuss if anything more needs to be done here! :)

**Comment by Grandolf49** [comment]
Hi @vorburger  Any more changes to be done here?

**Comment by vorburger** [comment]
> Hi @vorburger Any more changes to be done here?

Let me merge this as-is, as a first step... I'll raise a follow-up PR for your review to propose changing from Joda to Java time, and comment on FINERACT-838 for possible future next steps.

**Review by vorburger — APPROVED** [review]

**Pull request merged**